src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala

```diff
@@ -159,16 +186,27 @@ object TemporalQueryUtil extends LazyLogging {
 
   /**
    * join two temporalquery dataframes
+   *
+   * Nota bene:
+   * Columns which occur in both data frames df1 and df2 will occur only once in the result frame with df1 precedence over df2 using coalesce.
    */
   private def temporalJoinImpl( df1:DataFrame, df2:DataFrame, keyCondition:Column, joinType:String = "inner" )(implicit ss:SparkSession, hc:TemporalQueryConfig) : DataFrame = {
     // history join
+    df1.orderBy("id",hc.fromColName).show(false)
+
     val df2ren = df2.withColumnRenamed(hc.fromColName,hc.fromColName2).withColumnRenamed(hc.toColName,hc.toColName2)
+    df2ren.orderBy("id",hc.fromColName2).show(false)
     val df_join = df1.join( df2ren, keyCondition and col(hc.fromColName)<=col(hc.toColName2) and col(hc.toColName)>=col(hc.fromColName2), joinType)
+    df_join.show(false)
     // select final schema
-    val df1Cols = df1.columns.diff(hc.technicalColNames)
-    val df2Cols = df2.columns.diff(df1Cols ++ hc.technicalColNames)
-    val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_)) :+ greatest(col(hc.fromColName), col(hc.fromColName2)).as(hc.fromColName) :+ least(col(hc.toColName), col(hc.toColName2)).as(hc.toColName)
-    df_join.select(selCols:_*)
+    val dfCommonColNames = df1.columns.intersect(df2.columns).diff(hc.technicalColNames)
```

Comment:
What is the improvement of this refactoring?

Reply:
With the old implementation, 2 test cases fail: temporalFullJoin_dfRight, temporalRightJoin_dfRight.

The reason is that the old implementation gives priority to the columns of df₁: common columns of df₁ and df₂ are selected from df₁ only as they are removed from df2.columns:

```
val df1Cols = df1.columns.diff(hc.technicalColNames)
val df2Cols = df2.columns.diff(df1Cols ++ hc.technicalColNames)

val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_)) :+ ...
```

Note that the common columns occur in df1Cols only, but not in df2Cols.

In case you want to use `temporalJoinImpl` for a right join you would need to do the opposite:
```
val df2Cols = df2.columns.diff(hc.technicalColNames)
val df1Cols = df1.columns.diff(df2Cols ++ hc.technicalColNames)

val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_))) :+ ...
```

And in case of a full join ?

Well, best is an almost symmetric implementation where we select the common columns using `coalesce` as done now.

Why do only these 2 test cases fail?

Because these are the only tests where an entity occurs in df₂ only: There is not record with `id=1` in df₁.

Reply:
Sounds good, thanks for the rich explanation.